[browser] single threaded diagnostic server

pavelsavara opened the pull request:
This is split from https://github.com/dotnet/runtime/pull/110818 to make code review easier

Enabled by `<WasmPerfTracing>true</WasmPerfTracing>` and also `<MetricsSupport>` and `<EventSourceSupport>` as necessary.

Single thread
- `ep_rt_thread_create` and `ep_rt_thread_mono_start_func`
    - **it will run short, non-blocking jobs on the browser event loop instead of on dedicated threads**
    - `streaming_loop_tick` refactored from `streaming_thread`
    - `server_loop_tick` from refactored `server_thread`
- `mono_wasm_ds_exec` to process on tick on event loop
- `mono_schedule_ds_job` will register callback into queue for above
- NOOP implementation of `ep_rt_wait_event_wait`
- new `FEATURE_PERFTRACING_DISABLE_THREADS` -> `PERFTRACING_DISABLE_THREADS`
- `ds_ipc_stream_factory_get_next_available_stream` only polls once and never blocks

Sampling
- disabled `sampling_thread` because we can't do "stop-the-world and scan other threads"
- new events `ep_rt_sample_profiler_enabled` and `ep_rt_sample_profiler_disabled`
- new `ep_rt_mono_sampling_provider_component_init`
   - install mono profiler instrumentation `MONO_PROFILER_CALL_INSTRUMENTATION_ENTER`
   - this needs to happen before any managed code is executed 
- it has negative performance impact when workload & re-linked with `FeaturePerfTracing`
   - this is just single condition per method enter (and later per safe-point)
   - we could introduce separate `FeaturePerfTracingSampling` to enable/disable this if necessary

Web socket
- `FEATURE_PERFTRACING_PAL_WS`
- `ds-ipc-pal-websocket.c` and `ds-ipc-pal-websocket.h` they just forward to JavaScript
- empty for now `ds_rt_websocket_create`, `ds_rt_websocket_send`, `ds_rt_websocket_poll`, `ds_rt_websocket_recv`, `ds_rt_websocket_close`

Other
- include `diagnostics_tracing-static.a` in in-tree builds
- `Microsoft.NETCore.App.Runtime.Mono.perftrace.**RID**` cleanup
- fix `SystemNative_GetCpuUtilization`
- fix crash dump file permissions for CI/helix

Next PRs
- the rest of the implementation in JavaScript, for WebSocket and JS client
- better sampling instrumentation via `MINT_SDB_SEQ_POINT` or similar
- tests

Out of scope
- pause runtime during startup
- multi-threaded EP on browser - this will need some cross thread marshaling

contributes to https://github.com/dotnet/runtime/issues/76316

pavelsavara committed:
wip

pavelsavara committed:
fix

pavelsavara committed:
fix

pavelsavara committed:
chmod a+rw

pavelsavara commented:
@lateralusX please review this first step

pavelsavara committed:
fix

pavelsavara commented on src/mono/browser/build/BrowserWasmApp.targets:
```diff
@@ -362,6 +363,8 @@
       <EmscriptenEnvVars Include="ENABLE_JS_INTEROP_BY_VALUE=0" Condition="'$(WasmEnableJsInteropByValue)' != 'true'" />
       <EmscriptenEnvVars Include="WASM_ENABLE_SIMD=1" Condition="'$(WasmEnableSIMD)' != 'false'" />
       <EmscriptenEnvVars Include="WASM_ENABLE_SIMD=0" Condition="'$(WasmEnableSIMD)' == 'false'" />
+      <EmscriptenEnvVars Include="FEATURE_PERFTRACING=1" Condition="'$(FeaturePerfTracing)' != 'false'" />
```
Nope, this is just passing the value (which doesn't have to be same as default) to emscripten as env variable.

maraf replied:
`Configuration=Debug` => `FeaturePerfTracing=true` => `FEATURE_PERFTRACING=1`
`Configuration=Release` => `FeaturePerfTracing=''` => `FEATURE_PERFTRACING=1`

What is the reason to set `FeaturePerfTracing` based on `Configuration`?

pavelsavara replied:
Thank you!

pavelsavara replied:
CoreCLR includes EP always. In Browser, I would like to include it when workload+Debug. Unless user specified `FeaturePerfTracing` explicitly. I missed the Release -> empty branch. Fixed now.

lateralusX replied:
In Mono we have components meaning that an app can either statically or dynamically enable parts of the runtime when linking the app, using the same runtime build. EventPipe uses this (hot reload, debugger as well) and our mobile platforms uses it to enable/disable EventPipe using msbuild flags when building the app. Recommendation is to exclude it for release builds that should be pushed to app stores, but include it for debug and profiling builds of the apps.

pavelsavara commented on src/mono/System.Private.CoreLib/System.Private.CoreLib.csproj:
```diff
@@ -135,7 +135,7 @@
     <ILLinkDescriptorsXmls Include="$(ILLinkDirectory)ILLink.Descriptors.OSX.xml" Condition="'$(TargetsOSX)' == 'true' or '$(TargetsMacCatalyst)' == 'true' or '$(TargetsiOS)' == 'true' or '$(TargetstvOS)' == 'true'" />
     <ILLinkDescriptorsXmls Include="$(CoreLibSharedDir)ILLink\ILLink.Descriptors.Shared.xml" />
-    <ILLinkDescriptorsXmls Condition="'$(FeaturePerfTracing)' == 'true'" Include="$(CoreLibSharedDir)ILLink\ILLink.Descriptors.EventSource.xml" />
+    <ILLinkDescriptorsLibraryBuildXml Condition="'$(FeaturePerfTracing)' == 'true'" Include="$(CoreLibSharedDir)ILLink\ILLink.Descriptors.EventSource.xml" />
```
I wonder what is rooting it for x64 NAOT, when trimmed

lateralusX replied:
This make sure we don't link out the InitializeDefaultEventSources method called during Mono startup to setup runtime specific event sources. Mono root this function through above file and CoreClr does a similar thing, but CoreClr uses additional tasks to root methods called by runtime by scanning different source files, like corlib.h and generate a ILLink.Descriptors.xml as part of build. NativeAOT uses ModuleInitializer preventing this function from being linked out when running ilc on the published app but not exactly sure how its rooting it as part of S.P.C build, but I assume it ends up with similar mechanism.

lateralusX replied:
btw, why do we need to change fro ILLinkDescriptorsXmls to ILLinkDescriptorsLibraryBuildXml for just his specific item?

pavelsavara replied:
We run ILLink again during publish build step on customer dev machine. This file needs to be there to protect it from trimming. With ILLinkDescriptorsXmls it didn't work for me. Is this the right way ?

lateralusX replied:
For publish on dev machines, EventSource related logic is kept using linker flags, EventSourceSupport. That is how it works on all other platforms so I assume it should work the same on WASM

pavelsavara replied:
I will revert this change and fix it in later PR if necessary

pavelsavara committed:
Merge branch 'main' into browser_st_ep_step1

pavelsavara committed:
revert chmod

pavelsavara committed:
put it back

pavelsavara committed:
feedback

pavelsavara committed:
Merge branch 'main' into browser_st_ep_step1

lateralusX commented on src/native/eventpipe/ep-session.c:
```diff
@@ -34,6 +34,19 @@ ep_session_remove_dangling_session_states (EventPipeSession *session);
+	}
+	if (!ep_session_write_all_buffers_to_file (session, events_written)) {
+		session->streaming_thread = NULL;
+		ep_disable ((EventPipeSessionID)session);
```
I think we need to preserve this call to happen outside the EP_GCX_PREEMP_ENTER/EXIT as it did before.

pavelsavara replied:
Could you please explain why ?

lateralusX replied:
ep_disable can run managed code so thread needs to be back in coop mode before running that call.

lateralusX commented on src/native/eventpipe/ep-session.c:
```diff
@@ -34,6 +34,19 @@ ep_session_remove_dangling_session_states (EventPipeSession *session);
+	return 0; // try again
+}
+
 EP_RT_DEFINE_THREAD_FUNC (streaming_thread)
```
Maybe it is simpler and more clear to just create two functions keeping the first error/state checking code and then call into another function implemented differently for single and multi threaded support, you can still call the loop tick, but it might just be as simple to inline that part into each implementation once you have two methods.

lateralusX reviewed (commented):
Mainly looked at EventPipe and Mono native code in this first review pass.

pavelsavara commented on src/mono/mono/eventpipe/ep-rt-mono-runtime-provider.c:
```diff
@@ -1306,6 +1313,94 @@ ep_rt_mono_sample_profiler_write_sampling_event_for_threads (
+}
+
+static void
+method_enter (MonoProfiler *prof, MonoMethod *method, MonoProfilerCallContext *ctx)
```
That's right, it needs to consider sample frequency. But also I will bring another instrumentation for like safepoint.
I will make those improvements in next PR

lateralusX replied:
OK, if we ignore it now, it's important that it makes it into later PR's.

pavelsavara commented on src/mono/browser/browser.proj:
```diff
@@ -424,6 +424,8 @@
       <_CmakeEnvironmentVariable Include="ENABLE_JS_INTEROP_BY_VALUE=0" Condition="'$(WasmEnableJsInteropByValue)' == 'false'"/>
       <_CmakeEnvironmentVariable Include="WASM_ENABLE_SIMD=1" Condition="'$(WasmEnableSIMD)' != 'false'" />
       <_CmakeEnvironmentVariable Include="WASM_ENABLE_SIMD=0" Condition="'$(WasmEnableSIMD)' == 'false'" />
+      <_CmakeEnvironmentVariable Include="FEATURE_PERFTRACING=1" Condition="'$(FeaturePerfTracing)' != 'false'" />
```
Maybe there is same problem that Maraf reported

pavelsavara commented on src/mono/browser/browser.proj:
```diff
@@ -568,6 +570,8 @@
       <_MonoRollupEnvironmentVariable Include="WasmEnableThreads:$(WasmEnableThreads)" />
       <_MonoRollupEnvironmentVariable Include="WASM_ENABLE_SIMD:1" Condition="'$(WasmEnableSIMD)' != 'false'" />
       <_MonoRollupEnvironmentVariable Include="WASM_ENABLE_SIMD:0" Condition="'$(WasmEnableSIMD)' == 'false'" />
+      <_MonoRollupEnvironmentVariable Include="FEATURE_PERFTRACING:1" Condition="'$(FeaturePerfTracing)' != 'false'" />
```
Maybe there is same problem that Maraf reported

pavelsavara commented on src/mono/mono/component/CMakeLists.txt:
```diff
@@ -216,6 +216,15 @@ endif()
+
+    if (DISABLE_THREADS)
+        set_target_properties(
+            eventpipe-mono-objects
```
Here we are passing it to `src\native\eventpipe\CMakeLists.txt` so that it could set `FEATURE_PERFTRACING_SINGLE_THREADED`

lateralusX replied:
OK, why can't we just pass that cmake define directly instead of using DISABLE_THREADS? DISABLE_THREADS looks mono specific and probably shouldn't be used in generic eventpipe CMakeLists.txt.

pavelsavara commented on src/mono/mono/eventpipe/ep-rt-mono-runtime-provider.c:
```diff
@@ -25,8 +25,14 @@ extern EVENTPIPE_TRACE_CONTEXT MICROSOFT_WINDOWS_DOTNETRUNTIME_RUNDOWN_PROVIDER_
 #define NUM_NANOSECONDS_IN_1_MS 1000000
 
 // Sample profiler.
+#if defined(PERFTRACING_MULTI_THREADED)
```
you prefer `#if !defined(PERFTRACING_SINGLE_THREADED)` everywhere, right ?

lateralusX replied:
I think "multi threaded" should be the default, not sure how that is best expressed so it looks most intuitive, maybe we should pick up the use of DISABLE_THREADS instead of having single/multi threaded like we do in Mono, so PERFTRACING_DISABLE_THREADS and then only use that for checks. In case you just check one define you could do
#ifdef, #ifndef. Maybe we should primarily check for the exception (no threads) since multithreaded is default, so in that case we would mainly do:

#ifdef  PERFTRACING_DISABLE_THREADS
// Special case when we don't have threads.
#else
// Normal scenario.
#endif

What do you think? Most important is that it reads out well in the code and we should only end up with one define for this scenario and it should be the exception from default multithreaded behavior.

pavelsavara replied:
Ok, adopting  `FEATURE_PERFTRACING_DISABLE_THREADS` -> `PERFTRACING_DISABLE_THREADS` and drop all others

pavelsavara commented on src/mono/mono/eventpipe/ep-rt-mono-runtime-provider.c:
```diff
@@ -1306,6 +1313,94 @@ ep_rt_mono_sample_profiler_write_sampling_event_for_threads (
+#else // !PERFTRACING_MULTI_THREADED
+
+bool
+ep_rt_mono_sample_profiler_write_sampling_event_for_threads (
```
Correct, I just want to capture the `sampling_event` and `sampling_thread` to be able to use it later.

lateralusX replied:
OK, I had a comment embedded in some other comment that maybe we shouldn't "misuse" this function for this since its confusing just looking at the name, maybe make this into a noop since WASM does sampling differently and pass down needed data using the newly added enable function instead.

pavelsavara commented on src/mono/mono/eventpipe/ep-rt-mono-runtime-provider.c:
```diff
@@ -1306,6 +1313,94 @@ ep_rt_mono_sample_profiler_write_sampling_event_for_threads (
+method_filter (MonoProfiler *prof, MonoMethod *method)
+{
+	// TODO add more instrumentation, something like MINT_SDB_SEQ_POINT
+	return MONO_PROFILER_CALL_INSTRUMENTATION_ENTER;
```
Safepoint like instrumentation of interp sampling will be next PR. 
AOT should do something similar too but I have not investigated yet.
There is also WASM jiterp, which needs to make this work.

We can't run browser event loop in the middle of executing managed code.

lateralusX replied:
OK, so that means we can't rely on a timer or similar pattern putting new events on the event loop since we won't get back control while running managed code and we would need to sample at regular time interval while running managed code. That means we would need a very quick check when hitting sample points to decide if we are going to sample or not since it will be evaluated frequently. Safepoints only checks a global flag if it should call deeper into the safepoint mechanism, this global flag is controlled by a different thread requesting thread(s) to stop, but since WASM won't handle separate threads nothing external of the thread running managed code could check conditions and signal thread to stop and sample, meaning more logic needs to be baked into the sample points that would add overhead.

pavelsavara replied:
There is another concern related to events in synchronous managed code. Samples/events there could generate more data than is the size of the buffer. And "streaming thread" will not kick in, because there is no other thread. It didn't happen to me in my testing yet, but I'm sure it may happen with more CPU heavy workload.

I think the solution is to trigger `ep_session_write_all_buffers_to_file` from some location which has visibility into usage of that session buffer. 

Maybe in `ep_buffer_manager_write_event` instead of
```
	if (should_signal_reader_thread)
		// Indicate that there is new data to be read
		ep_rt_wait_event_set (&buffer_manager->rt_wait_event);
```

thoughts ? @lateralusX

lateralusX replied:
Lets look at that when we have the "safepoint" strategy in place, but I assume that we would need something that could trigger this without the need to have the streaming_loop_tick to run. Having that said, we would need to find a good place making sure we won't cause to frequent flushing of the internal buffers.

pavelsavara replied:
This is the instrumentation I would like to use for that https://github.com/dotnet/runtime/pull/112352
It's independent PR, we will need 3rd PR to integrate them together.

Re buffer flushing, that could be conditional, based of how full the buffer is.

lateralusX replied:
jupp, it would need to be a balance on how frequent you should get data into the stream as well as good use of allocated buffer.

pavelsavara commented on src/mono/mono/eventpipe/ep-rt-mono-runtime-provider.c:
```diff
@@ -1306,6 +1313,94 @@ ep_rt_mono_sample_profiler_write_sampling_event_for_threads (
+	_ep_rt_mono_sampling_profiler_provider = mono_profiler_create (NULL);
+	// this has negative performance impact even when the EP client is not connected!
+	// but it has to be enabled before managed code starts running, because the instrumentation needs to be in place
+	mono_profiler_set_call_instrumentation_filter_callback (_ep_rt_mono_sampling_profiler_provider, method_filter);
```
This would mean that interp could do IL->IR without instrumentation in the meantime, right ? 
And you would not see those ever after in the sample.
This is why I set it up during startup, not later.

lateralusX replied:
You will still do init in component_init but add a fini method like the other providers do for cleanup, when doing cleanup EP is shutting down, meaning you won't get any new events. Alternative is just to add this code into existing ep_rt_mono_runtime_provider_component_init and do cleanup in ep_rt_mono_runtime_provider_fini, sampling event is part of that runtime provider anyways.

pavelsavara commented on src/native/eventpipe/ep-sample-profiler.c:
```diff
@@ -192,6 +194,9 @@ sample_profiler_enable (void)
+
+#if defined(PERFTRACING_MULTI_THREADED)
 		EP_ASSERT (!ep_rt_wait_event_is_valid (&_thread_shutdown_event));
 		ep_rt_wait_event_alloc (&_thread_shutdown_event, true, false);
```
There is only busy spin wait on browser, you can't really wait.

pavelsavara replied:
> Regardless we could just make the code work on both single and multithreaded

Probably not
- we need to enter from interp loop, rather than from here.
- we capture only one stack trace
- the code for that is different on current thread than scanning another thread

lateralusX replied:
OK, my comment was mainly around reducing ifdefing by either do the same call if it's a noop or move the specialized code into separate function implemented differently depending on threading support.

Also, would be nice to not misuse the ep_rt_sample_profiler_write_sampling_event_for_threads in this function (a little confusing). Maybe adding needed info into the new ep_rt_sample_profiler_enable call instead and make ep_rt_sample_profiler_write_sampling_event_for_threads a noop for WASM?

pavelsavara committed:
Merge branch 'main' into browser_st_ep_step1

pavelsavara committed:
feedback:

- use PERFTRACING_DISABLE_THREADS
- introduce ep_rt_event_loop_job_create instead of
- pass EventPipeEvent to ep_rt_sample_profiler_enabled, so that I don't have to abuse ep_rt_sample_profiler_write_sampling_event_for_threads
- rename mono_schedule_ds_job and implement self-schedule there
- feedback about FeaturePerfTracing
- mark more stuff EP_UNREACHABLE
- revert ILLinkDescriptorsLibraryBuildXml

pavelsavara committed:
fix

pavelsavara commented on src/mono/mono/eventpipe/ep-rt-mono.h:
```diff
@@ -861,7 +904,13 @@ ep_rt_thread_create (
 		thread_params->thread_params.thread_func = (ep_rt_thread_start_func)thread_func;
 		thread_params->thread_params.thread_params = params;
 		thread_params->background_thread = true;
+#if defined(PERFTRACING_MULTI_THREADED)
```
I will create dedicated `ep_rt_event_loop_job_create` so that it's less confusing

pavelsavara committed:
fix

pavelsavara commented on src/mono/mono/eventpipe/ep-rt-mono.h:
```diff
@@ -714,7 +732,11 @@ ep_rt_wait_event_wait (
+#if defined(PERFTRACING_MULTI_THREADED)
 	return (int32_t)mono_w32handle_wait_one (wait_event->event, timeout, alertable);
+#else
+	return (int32_t)0;
```
NOOP

pavelsavara committed:
feedback: ep_rt_mono_sampling_provider_component_fini

pavelsavara commented:
> Mainly looked at EventPipe and Mono native code in this first review pass.

great feedback @lateralusX , many thanks!

I believe I processed all your feedback
- added `ep_rt_mono_sampling_provider_component_fini` to clean up the sampling/profiler
- fixed CMakeLists.txt and `FEATURE_PERFTRACING_DISABLE_THREADS` and `PERFTRACING_DISABLE_THREADS` 
- introduce `ep_rt_event_loop_job_create` instead of abusing `ep_rt_thread_create`
- pass `EventPipeEvent*` to `ep_rt_sample_profiler_enabled`, so that I don't have to abuse `ep_rt_sample_profiler_write_sampling_event_for_threads`
- rename `mono_schedule_ds_job` and implement self-schedule there
- marked more stuff EP_UNREACHABLE
- feedback about FeaturePerfTracing from Maraf
- reverted ILLinkDescriptorsLibraryBuildXml

I skipped solving sampling on safepoint and frequency, which will be in future PR.
I have new question WRT [flushing buffers](https://github.com/dotnet/runtime/pull/111910#discussion_r1941755403)

pavelsavara committed:
revert

pavelsavara committed:
linking modules

pavelsavara committed:
Merge branch 'main' into browser_st_ep_step1

# Conflicts:
#	src/native/eventpipe/CMakeLists.txt

lateralusX reviewed (commented):
Second pass on the EP code, mainly code structural comments as well as one idea around ep_rt_thread_create.

pavelsavara commented on src/native/eventpipe/configure.cmake:
```diff
@@ -17,6 +17,15 @@ if (CLR_CMAKE_HOST_IOS OR CLR_CMAKE_HOST_TVOS OR CLR_CMAKE_HOST_ANDROID)
   set(FEATURE_PERFTRACING_DISABLE_DEFAULT_LISTEN_PORT 1)
 endif()
 
+# Use WebSocket for EventPipe on browser
```
I just follow example few lines above where `CLR_CMAKE_HOST_IOS` does the same. Why is that different ?

lateralusX replied:
It's there to build for Native AOT and CoreClr, Mono cmake defines are handled through mono.proj.

lateralusX replied:
Not sure why it was put there instead of corresponding Native AOT or CoreClr specific cmake configuration files like we do on Mono.

pavelsavara commented on src/native/eventpipe/ds-server.c:
```diff
@@ -112,6 +112,63 @@ server_warning_callback (
+
+	ds_ipc_message_fini (&message);
+
+	(void)data; // unused
```
Yes, this is just to match the signature of job queue.

pavelsavara commented on src/native/eventpipe/ds-server.c:
```diff
@@ -218,6 +227,7 @@ ds_server_init (void)
+#ifndef PERFTRACING_DISABLE_THREADS
 		ep_rt_thread_id_t thread_id = ep_rt_uint64_t_to_thread_id_t (0);
 
 		if (!ep_rt_thread_create ((void *)server_thread, NULL, EP_THREAD_TYPE_SERVER, (void *)&thread_id)) {
```
We had that before.
It would have to allocate extra `rt_mono_thread_params_internal_t` which the job doesn't need. To keep the signatures same. But that's not main reason I reworked it. 
It was confusing and not easy to read. I like this much better.

lateralusX replied:
I was more thinking that you had two implementations of ep_rt_thread_create, one when you have thread support and one without, then there is no need to allocate rt_mono_thread_params_internal_t since you have a different implementation. That would remove all idefs around ep_rt_thread_create calls, so that part would be clearer, but it was just an idea, if you don't like it, that's fine.

pavelsavara committed:
Merge branch 'main' into browser_st_ep_step1

pavelsavara committed:
feedback

pavelsavara committed:
feedback

pavelsavara commented on src/mono/mono/eventpipe/ep-rt-mono.h:
```diff
@@ -865,6 +911,43 @@ ep_rt_thread_create (
+
+static
+bool
+ep_rt_event_loop_job_create (
```
I only renamed it to `ep_rt_queue_job`. 
Adding a structs feels like overkill. It would force itself into the `server_loop_tick` parameter.

lateralusX reviewed (approved):
Native Mono and EventPipe changes LGTM! Thanks!

pavelsavara committed:
Merge branch 'main' into browser_st_ep_step1

pavelsavara commented:
To be merged after preview snap
